### ISSUE TITLE
Error handling - no output from .on('error')

### ISSUE THREAD

**Issue by rosskevin** [adametry/gulp-eslint#135]
I would like to plugin my custom error handling/notifier.  In my other recipes such as [sass](https://github.com/alienfast/gulp-pipeline/blob/master/src/sass.js#L59), I simply use:
```javascript
.on('error', (error) => {
    this.notifyError(error, watching)
})
```
I see this plugin uses `throw Error` instead of a  `stream.emit('error')` [similar to scssLint](https://github.com/juanfran/gulp-scss-lint/blob/master/src/index.js#L36)

In order to work around this issue where `#failAfterError` doesn't emit an error to my handler, I [implemented a very ugly `#result` handler](https://github.com/alienfast/gulp-pipeline/blob/master/src/eslint.js#L56) mostly duplicating `#failAfterError` but it will trigger my own notifier (`#results` fails to notify).  

Note: this custom error handling attempts led me to discover to #134 

I'm new to gulp, but it seems like I should be able to listen for an error based on patterns of use from other plugins.  Is there any recommendation to improve [this code](https://github.com/alienfast/gulp-pipeline/blob/master/src/eslint.js#L56)?  It seems like an unnecessary hack.

**Comment by adametry** [comment]
*Note: I'll get into a bit of detail here. So, if you just want a solution skip to the workarounds.*

Would I be right to assume that you're referring to the `throw` in [failOnError](https://github.com/adametry/gulp-eslint/blob/master/index.js#L137) or [failAfterError](https://github.com/adametry/gulp-eslint/blob/master/index.js#L159)?
If so, this takes advantage of a convenience provided the `result` and `results` methods (respectively). Both use the `tryResultAction` method to catch thrown errors and pass them to the stream (the stream's onwrite function [emits the error](https://github.com/nodejs/readable-stream/blob/master/lib/_stream_writable.js#L365)).

I believe some confusion may be *when* the error is thrown. The `failAfterError` method creates a `results` (plural) stream, which waits until it has ended (technically, the "prefinish" event) to perform it's action once (collectively). Alternatively, the `failOnError` method creates a `result` (singular) stream, which will act (immediately) on each chunk/file. Likewise, the `format` method uses `results`, and the `formatEach` method uses `result`.

In your code, you work around the issue of `failAfterError` not behaving as expected by creating a `result` stream and use logic from the `failAfterError` method. In this case, you could simply use `failOnError` for the same effect; however, is likely end the stream before `format` can print lint results. You can simply use `failAfterError` in place of the custom `result` stream to emit an error after linting & printing all files. 

Now, to the heart of the matter: no "error" event. It appears that the issue may be that `gulp-if` is muting the error emitted from `failAfterError`. As you demonstrate in your code, throwing in the `result` method results in an "error" event.

**Three simple workarounds:**

1. Set the "error" event listener on the `failAfterError` stream (rather than the stream that `gulpIf` returns)
2. Remove `gulpIf`, use `results` instead of `result`, and add `!watching` to its failure condition.
3. Remove `gulpIf` and pipe to `failAfterError` with the "error" listener only if `!watching` (be sure to return the piped stream).

On a side note, you can throw anything from a `result` or `results` method. gulp-eslint will wrap handle wrapping the thrown value in a PluginError.

**Comment by rosskevin** [comment]
> Would I be right to assume that you're referring to the throw in failOnError or failAfterError?

Yes, with preferable use of `failAfterError`

> I believe some confusion may be when the error is thrown. 

Yes!  I am certainly confused.

> Now, to the heart of the matter: no "error" event. It appears that the issue may be that gulp-if is muting the error emitted

No, unfortunately not.  Shown below without `gulp-if` and the same results, my error listener is not called.
```javascript
      // 2. Test  https://github.com/adametry/gulp-eslint/issues/135#issuecomment-180109056
    .pipe(eslint.failAfterError()) 
    .on('error', (error) => {
      this.notifyError(error, watching)
    })
```

> As you demonstrate in your code, throwing in the result method results in an "error" event.

I believe this is an order of execution issue (which I do not quite grasp yet - a gulp thing).  I want `failAfterError` behavior, but have tripped upon the only solution being the `#result` method to trigger my `.on('error')`  I expect the code above to work, yet it doesn't.

Does that clarify?

**Comment by adametry** [comment]
It looks like you're not waiting around for the end of the stream. See the explanation I provided in issue #134 to understand why.

**Comment by rosskevin** [comment]
So, now that I've fixed the return of the stream, here are my results:

Error listener is not called (using gulpif):
```javascript
      // 2. Attempt now that returns are in place with the gulpif
      .pipe(gulpif(!watching, eslint.failAfterError())) 
      .on('error', (error) => {
        this.notifyError(error, watching)
      })
```

Error listener is not called (without the gulpif):
```javascript
      // 3. Attempt now that returns are in place WITHOUT gulpif
      .pipe( eslint.failAfterError()) 
      .on('error', (error) => {
        this.notifyError(error, watching)
      })
```

If I understand correctly, `.pipe( eslint.failAfterError()) ` should not continue until after all results are in, correct?

**Comment by adametry** [comment]
Indeed, `failAfterError` must "finish" (input closed event) and/or "end" (output closed event) before gulp will consider it done. For clarity, all files may pass through one or all streams in a pipeline before any stream ends.

How it works: `failAfterError` creates a pass-through stream that collects the sum of ESLint result errors. After the last file passes through, the writable (input) side of the stream will be closed (by whatever is piping into it). It's just before the "finish" event that `failAfterError` will throw (+ emit) an error if the error sum is greater than zero. This means that the error occurs after the last file has passed through, but before the `failAfterError` stream has emitted a "finish" or "end" event (ending the gulp task).

Can you place the following modified `failAfterError` method in place of `eslint.failAfterError()` in either example?
```javascript
	eslint.results(function(results) {
		var count = results.errorCount;
		console.log('Total ESLint Error Count: ' + count);
		if (count > 0) {
			throw new Error('Failed with Errors');
		}
	})
	.on('error', function(error) {
		console.log('Total ESLint Error Count: ' + error);
	})
	.on('finish', function() {
		console.log('eslint.results finished');
	})
	.on('end', function() {
		console.log('eslint.results ended');
	});
```
The console output should shed a bit of light on when the stream is closing and why there's no error.

**Comment by rosskevin** [comment]
Here's that log (had to add `pipe` to your sample):

```bash
[15:17:41] Requiring external module babel-core/register
[15:17:42] Using gulpfile ~/projects/acme/gulpfile.babel.js
[15:17:42] Starting 'eslint'...
[15:17:44] 
app/assets/javascripts/fake_analytics.js
  27:5  error  Expected a "break" statement before "case"  no-fallthrough
  29:5  error  Expected a "break" statement before "case"  no-fallthrough
  31:5  error  Expected a "break" statement before "case"  no-fallthrough
  33:5  error  Expected a "break" statement before "case"  no-fallthrough
  35:5  error  Expected a "break" statement before "case"  no-fallthrough

✖ 6 problems (6 errors, 0 warnings)

Total ESLint Error Count: 6
Total ESLint Error Count: Error in plugin 'gulp-eslint'
Message:
    Failed with Errors
Stack:
Error: Failed with Errors
    at Object.<anonymous> (eslint.js:92:17)
    at Object.exports.tryResultAction (/Users/kross/projects/gulp-pipeline/node_modules/gulp-eslint/util.js:137:11)
    at Transform._flush (/Users/kross/projects/gulp-pipeline/node_modules/gulp-eslint/index.js:124:8)
    at Transform.<anonymous> (_stream_transform.js:118:12)
    at Transform.g (events.js:260:16)
    at emitNone (events.js:67:13)
    at Transform.emit (events.js:166:7)
    at prefinish (_stream_writable.js:458:12)
    at finishMaybe (_stream_writable.js:466:7)
    at endWritable (_stream_writable.js:478:3)
    at Transform.Writable.end (_stream_writable.js:443:5)
    at Transform.onend (_stream_readable.js:490:10)
    at Transform.g (events.js:260:16)
    at emitNone (events.js:72:20)
    at Transform.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:905:12)
[15:17:44] 'eslint' errored after 1.92 s
[15:17:44] Error in plugin 'gulp-eslint'
Message:
    Failed with Errors
Stack:
Error: Failed with Errors
    at Object.<anonymous> (eslint.js:92:17)
    at Object.exports.tryResultAction (/Users/kross/projects/gulp-pipeline/node_modules/gulp-eslint/util.js:137:11)
    at Transform._flush (/Users/kross/projects/gulp-pipeline/node_modules/gulp-eslint/index.js:124:8)
    at Transform.<anonymous> (_stream_transform.js:118:12)
    at Transform.g (events.js:260:16)
    at emitNone (events.js:67:13)
    at Transform.emit (events.js:166:7)
    at prefinish (_stream_writable.js:458:12)
    at finishMaybe (_stream_writable.js:466:7)
    at endWritable (_stream_writable.js:478:3)
    at Transform.Writable.end (_stream_writable.js:443:5)
    at Transform.onend (_stream_readable.js:490:10)
    at Transform.g (events.js:260:16)
    at emitNone (events.js:72:20)
    at Transform.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:905:12)
eslint.results finished

Process finished with exit code 1
```

**Comment by rosskevin** [comment]
For clarity, the above was the result of:
```javascript
let bundle = this.gulp.src(this.config.source.glob, this.config.source.options)
      .pipe(gulpif(this.config.debug, debug(this.debugOptions())))
      .pipe(eslint(this.config.options))
      .pipe(eslint.format()) // outputs the lint results to the console. Alternatively use eslint.formatEach() (see Docs).

      // 4. https://github.com/adametry/gulp-eslint/issues/135#issuecomment-180555978
      .pipe(eslint.results(function (results) {
        var count = results.errorCount;
        console.log('Total ESLint Error Count: ' + count);
        if (count > 0) {
          throw new Error('Failed with Errors');
        }
      }))
      .on('error', function (error) {
        console.log('Total ESLint Error Count: ' + error);
      })
      .on('finish', function () {
        console.log('eslint.results finished');
      })
      .on('end', function () {
        console.log('eslint.results ended');
      })
```

**Comment by adametry** [comment]
_Oops. I should have changed the error log. No matter, it shows an emitted error._

The `[15:17:44] 'eslint' errored after 1.92 s` line and exit code suggest that gulp received the error thrown by the `eslint.results` stream. `failAfterError` does the same thing, only it populates a few more PluginError fields for gulp to report on. You should see a similar `'eslint' errored` line and subsequent error in your output with just `eslint.failAfterError()`.

I must be over looking something obvious, because it's essentially the same as:
```javascript
      .pipe( eslint.failAfterError()) 
      .on('error', (error) => {
        this.notifyError(error, watching)
      })
```
I'll run this through `gulp-pipeline` to see if I can find where the error is disappearing.

**Comment by rosskevin** [comment]
Ok, we are overlooking something:

7. WORKS! notification is emitted

```javascript
      // 7. notification is emitted
      .pipe(eslint.results(function (results) {
        let count = results.errorCount;
        if (count > 0) {
          throw new PluginError('gulp-eslint', { message: 'Failed with ' + count + (count === 1 ? ' error' : ' errors') })
        }
      }))
      .on('error', (error) => {
        this.notifyError(error, watching)
      })
```
8. Doesn't work (no notification emitted)

```javascript
      .pipe( eslint.failAfterError())
      .on('error', (error) => {
        this.notifyError(error, watching)
      })
```

I'm a bit mystified, because it's pretty clear that \#7 is pretty directly the impl of `#failAfterError`

**Comment by rosskevin** [comment]
Yikes, pulling my hair out.  While a standard run of 7 above works with a notification,  in watch I get no output from these:

```javascript
      .pipe(eslint.results(function (results) {
        let count = results.errorCount;
        if (count > 0) {
          console.error('****************in results handler')
          throw new PluginError('gulp-eslint', { message: 'Failed with ' + count + (count === 1 ? ' error' : ' errors') })
        }
      }))
      .on('error', (error) => {
        console.error('****************in error handler')
        this.notifyError(error, watching)
      })
```

**Comment by rosskevin** [comment]
BTW - this gulp-pipeline lint's itself, so if you clone, just run `gulp eslint` in that project and you are running that code.

**Comment by rosskevin** [comment]
This is a problem somewhere with gulp v3, I'm switching to gulp v4 and everything is working as expected, very strange indeed, but I'm going to close this as v4 is close to release and this probably isn't worth tracking down.
